### PR TITLE
SINGA-233 New communication interface for SINGA

### PR DESCRIPTION
**Pull request by cac2003**
In this ticket, we design and implement a new communication interface for SINGA. Specifically, each communication entity (physical node) is abstracted into an EndPoint, and users can directly call its send/recv method to accomplish data transfer.

### CODE REVIEW

**Comment by nudles** [comment]
pls update the commit msg (title and content).

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -9,9 +9,14 @@ IF(NOT USE_OPENCL)
     LIST(REMOVE_ITEM singa_test_source "singa/test_opencl.cc")
 ENDIF()
 
-ADD_EXECUTABLE(test_singa "gtest/gtest_main.cc" ${singa_test_source})
-ADD_DEPENDENCIES(test_singa singa_core singa_utils)
-MESSAGE(STATUS "link libs" ${singa_linker_libs})
-TARGET_LINK_LIBRARIES(test_singa gtest singa_core singa_utils singa_model
+#ADD_EXECUTABLE(test_singa "gtest/gtest_main.cc" ${singa_test_source})
```
Comment (nudles) [inline test/CMakeLists.txt]: pls uncomment these lines as they are necessary for compiling other tests.

File: test/singa/test_ep.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
+#include "singa/io/network/endpoint.h"
```
Comment (nudles) [inline test/singa/test_ep.cc]: pls add the license header to each new file.

File: include/singa/io/network/endpoint.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
+/************************************************************
+*
```
Comment (nudles) [inline include/singa/io/network/endpoint.h]: would it be better to include all header files of network into a single include/singa/io/network.h file, and put the .cc files into src/io/network/ folder?

File: include/singa/io/network/endpoint.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+#define MAX_RETRY_CNT 3
+
+#define EP_TIMEOUT 5.
+
+class NetworkThread;
+class EndPointFactory;
+
+class EndPoint {
+    private:
+        std::queue<Message*> send_;
```
Comment (nudles) [inline include/singa/io/network/endpoint.h]: we follow google coding style with 2space indent and max line width=80.
I can re-format the code.

File: include/singa/io/network/endpoint.h (hunk 3 above)
Comment (cac2003) [inline include/singa/io/network/endpoint.h]: fine with me. I will do that...

**Pull request merged**